server/metadata/tenant.go

```diff
@@ -609,85 +609,142 @@ func (tenant *Tenant) GetNamespace() Namespace {
 // metadata version once the commit is successful so reloading happens at the next call when a transaction sees a stale
 // tenant version. This applies to the reloading mechanism on all the servers. It returns "true" If the database already
 // exists, else "false" and the error.
-func (tenant *Tenant) CreateDatabase(ctx context.Context, tx transaction.Tx, dbName string, dbMetadata *DatabaseMetadata) (bool, error) {
+func (tenant *Tenant) CreateDatabase(ctx context.Context, tx transaction.Tx, dbName *DatabaseName, dbMetadata *DatabaseMetadata) error {
 	tenant.Lock()
 	defer tenant.Unlock()
 
-	if _, ok := tenant.databases[dbName]; ok {
-		return true, nil
+	key := dbName.Key()
+
+	if dbName.IsMainBranch() {
+		if _, ok := tenant.databases[key]; ok {
+			return NewDatabaseExistsErr(key)
+		}
+	} else {
+		// Create a database branch only if parent database exists
+		if _, ok := tenant.databases[dbName.Db()]; !ok {
+			return NewDatabaseNotFoundErr(dbName.Db())
+		}
+		if _, ok := tenant.databases[key]; ok {
+			return NewDatabaseBranchExistsErr(dbName.Branch())
+		}
 	}
 
 	// otherwise, proceed to create the database if there are concurrent requests on different workers then one of
 	// them will fail with duplicate entry and only one will succeed.
-	dbId, err := tenant.metaStore.CreateDatabase(ctx, tx, dbName, tenant.namespace.Id())
+	dbId, err := tenant.metaStore.CreateDatabase(ctx, tx, key, tenant.namespace.Id())
 	if dbMetadata != nil {
 		dbMetadata.SetDatabaseId(dbId)
-		err = tenant.namespaceStore.InsertDatabaseMetadata(ctx, tx, tenant.namespace.Id(), dbName, dbMetadata)
+		err = tenant.namespaceStore.InsertDatabaseMetadata(ctx, tx, tenant.namespace.Id(), key, dbMetadata)
 		if err != nil {
 			log.Err(err).Msg("Failed to insert database metadata")
-			return false, errors.Internal("Failed to setup db metadata")
+			return errors.Internal("Failed to setup db metadata")
 		}
 	}
-	return false, err
+	// TODO: copy schema if a new branch is being created
```

Comment:
What if instead of adding the logic in this method, we have a CreateBranch method on the tenant that do the same what you have here but also has a logic of copying collection schema, etc? Also, if we have separate APIs for Create/Delete branches then it is fine to have different methods in the tenant to handle branches.